File: QR Scanner/app/src/main/java/scanner/app/scan/qrcode/reader/adapter/GenerateAdapter.java
package scanner.app.scan.qrcode.reader.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;


import scanner.app.scan.qrcode.reader.R;
import scanner.app.scan.qrcode.reader.data.preference.GenerateModel;


public class GenerateAdapter extends RecyclerView.Adapter<GenerateAdapter.GenerateViewHolder> {

    Context context;
    List<GenerateModel> generateModelList;

    private ClickListener clickListener;


    @NonNull
    @Override
    public GenerateViewHolder onCreateViewHolder( ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.generate_recycler, parent, false);
        return new GenerateViewHolder(view, viewType);
    }

    @Override
    public void onBindViewHolder( GenerateViewHolder holder, int position) {

        holder.generateImageView.setImageResource(generateModelList.get(position).getImg_icon());
        holder.generateTextView.setText(generateModelList.get(position).getCategory_name());




    }

    public void setClickListener(ClickListener clickListener) {
        this.clickListener = clickListener;
    }

    @Override
    public int getItemCount() {
        return generateModelList.size();
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
    }


    public class GenerateViewHolder extends RecyclerView.ViewHolder{

        RelativeLayout generateRelative;
        ImageView generateImageView;
        TextView generateTextView;

        public GenerateViewHolder(View itemView,int viewType) {
            super(itemView);

            generateRelative=itemView.findViewById(R.id.anotherRecycler);
            generateImageView=itemView.findViewById(R.id.generate_txt_img);
            generateTextView=itemView.findViewById(R.id.generate_text);

            generateRelative.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (clickListener != null) {
                        clickListener.onItemClicked(getLayoutPosition());
                    }
                }
            });

        }
    }

    public interface ClickListener {

         void onItemClicked(int position);

    }

    public GenerateAdapter(Context context, List<GenerateModel> generateModelList) {
        this.context = context;
        this.generateModelList = generateModelList;
    }
}

File: QR Scanner/app/src/main/java/com/infinity/interactive/scanqr/generateqr/adapter/GenerateAdapter.kt
package com.infinity.interactive.scanqr.generateqr.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.infinity.interactive.scanqr.generateqr.data.preference.GenerateModel
import com.infinity.interactive.scanqr.generateqr.databinding.GenerateRecyclerBinding

class GenerateAdapter(
    private val context: Context,
    private val generateModelList: List<GenerateModel>
) : RecyclerView.Adapter<GenerateAdapter.GenerateViewHolder>() {

    private var clickListener: ClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GenerateViewHolder {
        val binding =
            GenerateRecyclerBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return GenerateViewHolder(binding)
    }

    override fun onBindViewHolder(holder: GenerateViewHolder, position: Int) {
        holder.bind(generateModelList[position])
    }

    fun setClickListener(clickListener: ClickListener?) {
        this.clickListener = clickListener
    }

    override fun getItemCount(): Int {
        return generateModelList.size
    }

    inner class GenerateViewHolder(private val binding: GenerateRecyclerBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(generateModel: GenerateModel) {

            binding.generateTxtImg.setImageResource(generateModel.img_icon)
            binding.generateText.text = generateModel.category_name
            binding.anotherRecycler.setOnClickListener {
                clickListener?.onItemClicked(layoutPosition)
            }
        }
    }

    interface ClickListener {
        fun onItemClicked(position: Int)
    }
}

